File: Week 4/cpp/5.alpha_pattern.cpp
/*
 * WEEK 4 - C++ PATTERN PROBLEMS
 * Topic: Alpha Pattern
 * File: 5.alpha_pattern.cpp
 *
 * PATTERN (N=4):
 *  A
 *  BB
 *  CCC
 *  DDDD
 *
 * KEY POINTS:
 *  - char arithmetic in C++ is implicit: 'A' + 1 yields int 66.
 *  - Cast back with static_cast<char>(...) to print as a letter.
 */

#include <iostream>
#include <string>

int main() {
    int n;
    if (!(std::cin >> n)) { n = 4; std::cout << "(demo n=4)\n"; }
    for (int i = 1; i <= n; ++i) {
        char ch = static_cast<char>('A' + i - 1);
        for (int j = 1; j <= i; ++j) std::cout << ch;
        std::cout << '\n';
    }
    std::cout << "--- string variant ---\n";
    for (int i = 1; i <= n; ++i)
        std::cout << std::string(i, static_cast<char>('A' + i - 1)) << '\n';
    return 0;
}

/*
 * NOTES:
 *  - 'A' + i - 1 produces an int; cast to char before output.
 *  - For N > 26 you'd run off 'Z' -- consider modulo if wrapping is desired.
 */

File: Week 4/python/5.alpha_pattern.py
"""
WEEK 4 - PYTHON PATTERN PROBLEMS
Topic: Alpha Pattern (single character per row)
File: 5.alpha_pattern.py

PATTERN (N=4):
A
BB
CCC
DDDD

CONCEPT:
Row i prints the i-th letter of the alphabet, repeated i times. Map row
number to character via `chr(ord('A') + i - 1)`.

KEY POINTS:
 - `chr(n)` -> character with code n.
 - `ord(ch)` -> integer code of character ch.
 - Pythonic: `print(chr(64 + i) * i)`.

DRY RUN (N=3):
 row 1 -> 'A' -> "A"
 row 2 -> 'B' -> "BB"
 row 3 -> 'C' -> "CCC"

COMPLEXITY: O(N^2).
"""

import sys


def main() -> None:
    n = 4 if sys.stdin.isatty() else int(sys.stdin.read().split()[0])

    for i in range(1, n + 1):
        ch = chr(ord('A') + i - 1)
        for _ in range(i):
            print(ch, end='')
        print()
    print("--- pythonic ---")
    for i in range(1, n + 1):
        print(chr(ord('A') + i - 1) * i)


if __name__ == "__main__":
    main()


# NOTES:
# - For row > 26 you'd run off the alphabet -- add `% 26` if you want to wrap.
# - In Python `'A' + 1` is a TypeError; you MUST use ord/chr explicitly.
